src/NHibernate/Loader/Loader.cs

```diff
@@ -1637,48 +1637,73 @@ protected IList List(ISessionImplementor session, QueryParameters queryParameter
 		/// <returns></returns>
 		protected IList List(ISessionImplementor session, QueryParameters queryParameters, ISet<string> querySpaces)
 		{
-			bool cacheable = _factory.Settings.IsQueryCacheEnabled && queryParameters.Cacheable;
-
-			if (cacheable)
-			{
-				return ListUsingQueryCache(session, queryParameters, querySpaces);
-			}
-			return ListIgnoreQueryCache(session, queryParameters);
+			return ListUsingQueryCacheOrNull(session, queryParameters, querySpaces)
+					?? ListIgnoreQueryCache(session, queryParameters);
 		}
 
 		private IList ListIgnoreQueryCache(ISessionImplementor session, QueryParameters queryParameters)
 		{
 			return GetResultList(DoList(session, queryParameters), queryParameters.ResultTransformer);
 		}
 
-		private IList ListUsingQueryCache(ISessionImplementor session, QueryParameters queryParameters, ISet<string> querySpaces)
+		internal void ProcessCachedResults(
+			ISessionImplementor session,
+			QueryParameters queryParameters,
+			ISet<string> querySpaces,
+			Action<IQueryCache, QueryKey, IList> cacheProcessingLogic)
 		{
-			IQueryCache queryCache = _factory.GetQueryCache(queryParameters.CacheRegion);
+			bool cacheable = _factory.Settings.IsQueryCacheEnabled && queryParameters.Cacheable;
 
-			QueryKey key = GenerateQueryKey(session, queryParameters);
+			if (!cacheable)
+				return;
+
+			var queryCache = _factory.GetQueryCache(queryParameters.CacheRegion);
+			var key = GenerateQueryKey(session, queryParameters);
 
 			IList result = GetResultFromQueryCache(session, queryParameters, querySpaces, queryCache, key);
+			cacheProcessingLogic(queryCache, key, result);
+		}
 
-			if (result == null)
-			{
-				result = DoList(session, queryParameters, key.ResultTransformer);
-				PutResultInQueryCache(session, queryParameters, queryCache, key, result);
-			}
+		private IList ListUsingQueryCacheOrNull(ISessionImplementor session, QueryParameters queryParameters, ISet<string> querySpaces)
+		{
+			IList cachedResults = null;
+			ProcessCachedResults(
+				session,
+				queryParameters,
+				querySpaces,
+				(queryCache, cacheKey, results) =>
+				{
+					if (results == null)
+					{
+						results = DoList(session, queryParameters, cacheKey.ResultTransformer);
+						PutResultInQueryCache(session, queryParameters, queryCache, cacheKey, results);
+					}
+
+					TransformCachedResults(queryParameters, cacheKey.ResultTransformer, ref results);
+					cachedResults = results;
+				});
 
+			//means cache is disabled
+			if (cachedResults == null)
+				return null;
+
+			return GetResultList(cachedResults, queryParameters.ResultTransformer);
+		}
+
+		internal void TransformCachedResults(QueryParameters queryParameters, CacheableResultTransformer transformer, ref IList result)
```

Comment:
`ref` and `out` parameters are to be avoided whenever possible. Having a `void` return type with a `ref` calls for refactoring the `ref` out of the method signature.